### DISCUSSION TITLE
allow some errors to propagate up in embedded usage

### DISCUSSION THREAD

**Discussion by mikerott** [projectdiscovery/nuclei#2287] [Ideas]
I'm using Nuclei embedded as a lib, where I provide my own implementation of the [output.Writer](https://github.com/projectdiscovery/nuclei/blob/master/v2/pkg/output/output.go#L30) to collect all of the `ResultEvent`s and `InternalEvent`s.

In the case where the hostname is not resolvable or reachable or times out, _no function_ in the `writer` is called at all.  In my embedded usecase, I'd very much like to see if the reason a particular host was not scanned.

As an example, errors are already returned from [the various Dial* functions in fastdialer](https://github.com/projectdiscovery/fastdialer/blob/master/fastdialer/dialer.go):
* [`NoAddressFoundError`](https://github.com/projectdiscovery/fastdialer/blob/master/fastdialer/dialer.go#L185)
* [`CouldNotConnectError`](https://github.com/projectdiscovery/fastdialer/blob/master/fastdialer/dialer.go#L260)

... but do not propagate up into the `output.Writer`.  It would help me to know if the reason a host lacks an event is due to unresolvable, unreachable, timeout, etc.

I'm not sure the best way to achieve this.  Perhaps another `output.Writer` callback function?  Maybe include the error somewhere in the `InternalEvent` that is passed to the `output.Writer`'s `WriteFailure` callback function?

Thanks for considering!

**Comment by mikerott** [comment]
I opened a PR with a proposed change here:  https://github.com/projectdiscovery/nuclei/pull/2291

The PR allows a developer who is using Nuclei embedded to implement their own `hosterrorscache.CacheInterface` which has functions with sufficient signature and Nuclei usage to offer the developer an opportunity to capture host-related errors (like "no address found", "timed out", etc).

**Comment by dogancanbakir** [comment]
This discussion closed automatically due to inactivity. Feel free to reopen or start new if still relevant.
